### PR TITLE
Force student to type something before ending the exam

### PR DESCRIPTION
**Pull request by stuwilmur**
**Not for merging**: a starting point for further discussion and work.

**Background**
Work on implementing [Issue #515](https://github.com/numbas/Numbas/issues/515): Force student to type something before ending exam.

**Changes**
- Added new confirm end exam modal in default theme, with text input. When the specified text (case-insensitive) is entered, the OK button will end the exam.
- Added new function `showConfirmEndExam` in display-base.js, and call from exam.js

**For discussion**
- Should this feature be configurable, e.g. off by default? It might be annoying or unexpected for some.
- Should the input be required regardless of how many questions have been attempted/submitted?
- The current implementation only works for en-GB: until the “end” command is translated it will be impossible to end the exam in other languages!
- Test to be added

### CODE REVIEW

**Comment by christianp** [comment]
Thanks for this!

Answers to your bullet points:

> Should this feature be configurable, e.g. off by default? It might be annoying or unexpected for some.

Yes, it should be off by default. I think this is only worth having for high-stakes exams.

> Should the input be required regardless of how many questions have been attempted/submitted?

Yes.

> The current implementation only works for en-GB: until the “end” command is translated it will be impossible to end the exam in other languages!

It'll just look for "end" regardless of language, won't it?

> Test to be added

There are no unit tests for the display code yet, so this isn't a problem.

I'll use GitHub's review thingy to go through the code you've committed so far.

**Comment by stuwilmur** [comment]
Thank *you* for the review!

Most changes done; I've managed to break the data binding (I think) when I attempted to move to the correct viewmodel; would you mind giving me a hint to try to get it working again?

> Thanks for this!
> 
> Answers to your bullet points:
> 
> > Should this feature be configurable, e.g. off by default? It might be annoying or unexpected for some.
> 
> Yes, it should be off by default. I think this is only worth having for high-stakes exams.

Great: I will try to implement this on the next pass
> 
> > Should the input be required regardless of how many questions have been attempted/submitted?
> 
> Yes.

Thanks for confirming.
> 
> > The current implementation only works for en-GB: until the “end” command is translated it will be impossible to end the exam in other languages!
> 
> It'll just look for "end" regardless of language, won't it?

I need to check this again, once I get it working again (since I've managed to break the data binding)
> 
> > Test to be added
> 
> There are no unit tests for the display code yet, so this isn't a problem.
I thought that might be the case.
> 
> I'll use GitHub's review thingy to go through the code you've committed so far.

Thank you!

File: themes/default/templates/modals.html [hunk 1]
```diff
@@ -23,14 +25,44 @@ <h4 id="confirm-modal-title" data-localise="modal.confirm"></h4>
...
+            <div class="modal-header">
+                <h4 id="confirm-end-exam-modal-title" data-localise="modal.confirm"></h4>
+            </div>
+            <div class="modal-body" id="confirm-end-exam-modal-body">
+                <div class="modal-body" id="confirm-end-exam-modal-message"></div>
+                <form data-bind="submit: function(){}, with: $root.confirm_end">
+                    <div class="row">
+                        <div class="form-group col-sm-6">
+                            <label for="confirm-end-text-input" data-localise="modal.confirm-end-exam"></label>
+                            <input class="form-control" id="confirm-end--text-input" type="text"
```
[the remark this answers is not in the document]
Comment (stuwilmur) [reply themes/default/templates/modals.html]: Corrected

File: themes/default/templates/modals.html [hunk 2]
```diff
@@ -1,5 +1,6 @@
 <!-- modal dialog boxes -->
-<div class="modal fade" id="alert-modal" tabindex="-1" role="dialog" aria-hidden="true" aria-labelledby="alert-modal-title" aria-describedby="alert-modal-body">
+<div class="modal fade" id="alert-modal" tabindex="-1" role="dialog" aria-hidden="true"
+    aria-labelledby="alert-modal-title" aria-describedby="alert-modal-body">
```
[the remark this answers is not in the document]
Comment (stuwilmur) [reply themes/default/templates/modals.html]: Yes, reverted: sorry for the confusion.

File: themes/default/files/scripts/display-base.js [hunk 3]
```diff
@@ -220,10 +223,40 @@ var display = Numbas.display = /** @lends Numbas.display */ {
...
 
+    /** Compare two strings, ignoring differences in case.
+     * Does not ignore differences in accent, even where
+     * base characters are identical
+    *
+    * @param {string} a - reference string
+    * @param {string} b - comparison string
+    * @returns {boolean}
+    */
+    caselessCompare: function (a,b) {
```
[the remark this answers is not in the document]
Comment (stuwilmur) [reply themes/default/files/scripts/display-base.js]: Moved to `Numbas.util`; you're welcome!

File: themes/default/files/scripts/display-base.js [hunk 4]
```diff
@@ -220,10 +223,40 @@ var display = Numbas.display = /** @lends Numbas.display */ {
     showConfirm: function(msg,fnOK,fnCancel) {
         this.modal.ok = fnOK || function(){};
         this.modal.cancel = fnCancel || function(){};
-        $('#confirm-modal .modal-body').html(msg);
+        $('#confirm-modal.modal-body').html(msg);
```
[the remark this answers is not in the document]
Comment (stuwilmur) [reply themes/default/files/scripts/display-base.js]: Good catch: corrected.

File: tests/numbas-runtime.js [hunk 5]
```diff
@@ -24070,6 +24070,10 @@ Exam.prototype = /** @lends Numbas.Exam.prototype */ {
         }
```
[the remark this answers is not in the document]
Comment (stuwilmur) [reply tests/numbas-runtime.js]: I will try and remove this runtime file from the PR.

File: themes/default/files/scripts/display-base.js [hunk 6]
```diff
@@ -80,6 +80,9 @@ var display = Numbas.display = /** @lends Numbas.display */ {
             },
             staged_style: {
                 textSize: Knockout.observable('')
+            },
+            confirm_end: {
+                confirmation: Knockout.observable('')
```
[the remark this answers is not in the document]
Comment (stuwilmur) [reply themes/default/files/scripts/display-base.js]: Agreed: I've attempted to move this, but have either broken the data binding or am not referencing correctly from display-base.js. Would you mind pointing me in the right direction?

**Comment by stuwilmur** [comment]
Update: fixed the broken function reference.

**Comment by christianp** [comment]
Thanks for this. I've merged in the most recent commits, and made a change: I noticed that the password entry on the front page also asks for a string and gives feedback about whether it's correct, so that logic should be shared with the "confirm end" box.

I fixed a couple of things in the modal as well - there was a repeated `modal-body` class, and I made it so that submitting the form by pressing enter is the same as clicking the OK button.

Still to do:

* Maybe the "OK" button should be labelled "End exam".
* Add an exam setting to not require entering this string. The default should be not to require it.

**Comment by stuwilmur** [comment]
Implemented requested changes.
See [#801](https://github.com/numbas/editor/pull/801) for accompanying editor changes to support new option.

File: locales/en-GB.json [hunk 7]
```diff
@@ -19,6 +19,7 @@
     "extension.not found": "Couldn't load the extension <code>{{name}}</code>.",
     "modal.confirm": "Confirm",
     "modal.confirm end exam": "Write <code>{{endConfirmation}}</code> in the box to confirm:",
+    "modal.end exam button": "End exam",
```
Comment (christianp) [inline locales/en-GB.json:22]: Is there a reason you don't use the existing string `control.end exam`?

**Review by christianp — APPROVED** [review]
Works well, and I have no complaints about the code. Thanks!

**Pull request merged**